numba/unicode.py

```diff
@@ -1286,6 +1288,134 @@ def impl(a):
     return impl
 
 
+@overload_method(types.UnicodeType, 'islower')
+def unicode_islower(data):
+    """
+    impl is an approximate translation of:
+    https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/unicodeobject.c#L11900-L11933
+    mixed with:
+    https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/bytes_methods.c#L131-L156
+    """
+
+    def impl(data):
+        length = len(data)
+        if length == 1:
+            return _PyUnicode_IsLowercase(_get_code_point(data, 0))
+        if length == 0:
+            return False
+
+        cased = False
+        for idx in range(length):
+            code_point = _get_code_point(data, idx)
+            if _PyUnicode_IsUppercase(code_point) or _PyUnicode_IsTitlecase(code_point):
+                return False
+            elif not cased and _PyUnicode_IsLowercase(code_point):
+                cased = True
+        return cased
+    return impl
+
+
+@register_jitable
+def _handle_capital_sigma(data, length, idx):
+    """
+    Handle capital sigma
+    https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Objects/unicodeobject.c#L9856-L9883
+    """
+    c = 0
+    j = idx - 1
+    while j >= 0:
+        c = _get_code_point(data, j)
+        if not _PyUnicode_IsCaseIgnorable(c):
+            break
+        j -= 1
+    final_sigma = (j >= 0 and _PyUnicode_IsCased(c))
+    if final_sigma:
+        j = idx + 1
+        while j < 0:
```

Comment:
I am not sure this is correct. The cpython reads `for (j = i + 1; j < length; j++) {` so should `j` not be `< length`? If this isn't being caught by the unit tests,  I suggest to look at the unit tests themselves.

Reply:
Interesting side note, this is probably one of the few places where a `for` `else` loop is applicable, at least in Python:

```
In [7]: for i in range(0, 10):
   ...:     print(i)
   ...: else:
   ...:     i += 1
   ...:     print(i)
   ...:
   ...:
0
1
2
3
4
5
6
7
8
9
10

In [8]: i
Out[8]: 10
```

Reply:
Fixed and added test case for that.